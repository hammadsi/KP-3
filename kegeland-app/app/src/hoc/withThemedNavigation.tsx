import {NavigationContainer} from '@react-navigation/native';
import React, {useEffect, useState} from 'react';
import {Provider as PaperProvider} from 'react-native-paper';

import {darkTheme, lightTheme} from '~constants/theme';
import useAppSelector from '~hooks/useAppSelector';

const getTheme = (isDarkMode: boolean) => {
  return isDarkMode ? darkTheme : lightTheme;
};

/**
 * Higher-order component withThemedNavigation.
 * Wraps a component with a themed navigation container
 * @param Component the component to wrap
 */
const withThemedNavigation =
  <P extends object>(Component: React.FC<P>): React.FC<P> =>
  (props) => {
    const {darkMode} = useAppSelector((state) => state.app.settings);
    const [theme, setTheme] = useState(getTheme(darkMode));

    useEffect(() => {
      setTheme(getTheme(darkMode));
    }, [darkMode]);

    return (
      <PaperProvider theme={theme}>
        <NavigationContainer theme={theme}>
          <Component {...props} />
        </NavigationContainer>
      </PaperProvider>
    );
  };

export default withThemedNavigation;
