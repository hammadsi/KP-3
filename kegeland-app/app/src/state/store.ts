import {combineReducers, configureStore} from '@reduxjs/toolkit';
import AsyncStorage from '@react-native-async-storage/async-storage';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import reduxFlipper from 'redux-flipper';

import appReducer from './ducks/app/app.reducer';
import bluetoothReducer from './ducks/bluetooth/bluetooth.reducer';
import authReducer from './ducks/auth/auth.reducer';
import questionsReducer from './ducks/questions/questions.reducer';
import sessionReducer from './ducks/session/session.reducer';

import rootSaga, {sagaMiddleware} from './rootSaga';

/**
 * Configuration for redux-persist.
 * Redux-persist is responsible for maintaining the state
 * between sessions.
 */
const persistConfig = {
  key: 'root',
  version: 1,
  storage: AsyncStorage,
  whitelist: ['app', 'auth', 'bluetooth'],
};

// Initiate the root reducer
export const rootReducer = combineReducers({
  app: appReducer,
  auth: authReducer,
  bluetooth: bluetoothReducer,
  questions: questionsReducer,
  session: sessionReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

// Check if development-environment is enabled
const isDev = !['production', 'test', 'integration'].includes(
  process.env.NODE_ENV || '',
);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    isDev
      ? getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        })
          .concat(logger)
          .concat(reduxFlipper())
          .concat(sagaMiddleware)
      : getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }).concat(sagaMiddleware),
  devTools: isDev,
});

// Start redux-saga
sagaMiddleware.run(rootSaga);

// Create persistor for persisting state
export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
