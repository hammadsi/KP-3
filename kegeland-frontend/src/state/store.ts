import { combineReducers, configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import authReducer from './ducks/auth/auth.reducer';
import patientsReducer from './ducks/patients/patients.reducer';
import questionnairesReducer from './ducks/questionnaires/questionnaires.reducer';
import { sensorReducer } from './ducks/sensors/sensors.reducer';
import sessionsReducer from './ducks/sessions/sessions.reducer';
import settingsReducer from './ducks/settings/settings.reducer';
import wheelchairPatientsReducer from './ducks/wheelchairPatients/wheelchairPatients.reducer';

const persistConfig = {
  key: 'root',
  version: 1,
  storage,
  whitelist: ['settings', 'auth'],
};

export const rootReducer = combineReducers({
  settings: settingsReducer,
  auth: authReducer,
  patients: patientsReducer,
  sensors: sensorReducer,
  sessions: sessionsReducer,
  questionnaires: questionnairesReducer,
  wheelchairPatients: wheelchairPatientsReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const isDev = !['production', 'test'].includes(process.env.NODE_ENV || '');

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    isDev
      ? getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }).concat(logger)
      : getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }),
  devTools: isDev,
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
