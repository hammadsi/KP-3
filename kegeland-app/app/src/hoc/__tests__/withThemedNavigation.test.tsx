import React from 'react';
import {View} from 'react-native';
import {render} from '@testing-library/react-native';
import {Provider} from 'react-redux';

import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';
import withThemedNavigation from '~hoc/withThemedNavigation';

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe('Test withThemedNavigation-hoc', () => {
  const MockComponent: React.FC<any> = (props) => {
    return <View testID="MockComponent" {...props} />;
  };

  it('should render correctly', () => {
    const Component = withThemedNavigation(MockComponent);

    const component = render(
      <Provider store={mockStore(initialStore)}>
        <Component />
      </Provider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
