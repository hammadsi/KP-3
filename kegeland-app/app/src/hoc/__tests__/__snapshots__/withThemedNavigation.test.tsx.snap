// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test withThemedNavigation-hoc should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    testID="MockComponent"
  />
</View>
`;
